File: extension6/SupplyManager.py
from __future__ import annotations
from typing import TYPE_CHECKING, Dict
import time

if TYPE_CHECKING:
    from extension6.CoffeeHouseWithSupplyManager import CoffeeHouseWithSupplyManager
    from root.Ingredient import Ingredient
    from root.Product import Product


class SupplyManager:
    _model: CoffeeHouseWithSupplyManager
    _state: bool
    _ingredientInitialAmountMap: Dict[Ingredient, int]
    _productInitialAmountMap: Dict[Product, int]

    def __init__(self, model: CoffeeHouseWithSupplyManager):
        self._model = model
        self._state = True
        self._ingredientInitialAmountMap = dict()
        self._productInitialAmountMap = dict()

    def renewIngredient(self, ingredient: Ingredient) -> None:
        self._state = False
        print(f'Supply manager {self.__hash__()} left to get missing ingredient(Missing ingredient is {ingredient}) ')
        time.sleep(self._ingredientInitialAmountMap[ingredient]/10)
        self._model.addIngredient(ingredient, self._ingredientInitialAmountMap[ingredient])
        print(f'Supply manager {self.__hash__()} successfully bought missing ingredient({ingredient}).\nMissing '
              f'ingredient is now renewed')
        self._state = True

    def renewProduct(self, product: Product) -> None:
        self._state = False
        print(f'Supply manager {self.__hash__()} left to get missing product(Missing ingredient is {product}) ')
        time.sleep(self._productInitialAmountMap[product]/10)
        self._model.addDessert(product, self._productInitialAmountMap[product])
        print(f'Supply manager {self.__hash__()} successfully bought missing product({product}).\nMissing '
              f'product is now renewed')
        self._state = True

    def getManagerState(self) -> bool:
        return self._state

    def addIngredientInitialAmount(self, ingredient: Ingredient, amount: int) -> None:
        if ingredient not in self._ingredientInitialAmountMap:
            self._ingredientInitialAmountMap[ingredient] = amount

    def addProductInitialAmount(self, product: Product, amount: int) -> None:
        if product not in self._productInitialAmountMap:
            self._productInitialAmountMap[product] = amount

File: cplusplus/ext6/SupplyManager.cpp
#include "SupplyManager.hpp"

SupplyManager::SupplyManager(CoffeeHouseWithSupplyManager* model) : model(model), state(true)
{
}

long SupplyManager::renewIngredient(Ingredient* ingredient, long& currentTime)
{
	this->state = false;
	std::cout << "Time= " << currentTime << " Supply manager " << this->getHashCode() << " left to get missing ingredient(Missing ingredient is " << ingredient->getName() << ") " << std::endl;
	currentTime += this->ingredientInitialAmountMap[ingredient] / 10;
	this->model->addIngredient(ingredient, this->ingredientInitialAmountMap[ingredient]);
	std::cout << "Time= " << currentTime << " Supply manager " << this->getHashCode() << " successfully bought missing ingredient(" << ingredient->getName() << ")." << std::endl;
	std::cout << "Missing ingredient is now renewed";
	this->state = true;
	return currentTime;
}

long SupplyManager::renewProduct(Product* product, long& currentTime)
{
	this->state = false;
	std::cout << "Time= " << currentTime << " Supply manager " << this->getHashCode() << " left to get missing product(Missing ingredient is " << product->toString() << ") " << std::endl;
	currentTime += this->productInitialAmountMap[product] / 10;
	this->model->addDessert(product, this->productInitialAmountMap[product]);
	std::cout << "Time= " << currentTime << " Supply manager " << product->getHashCode() << " successfully bought missing product(" << product->toString() << ")." << std::endl;
	std::cout << "Missing product is now renewed";
	this->state = true;
	return currentTime;
}

bool SupplyManager::getManagerState()
{
	return this->state;
}

void SupplyManager::addIngredientInitialAmount(Ingredient* ingredient, int amount)
{
	auto it = this->ingredientInitialAmountMap.find(ingredient);
	if (it == this->ingredientInitialAmountMap.end())
		this->ingredientInitialAmountMap[ingredient] = amount;
}

void SupplyManager::addProductInitialAmount(Product* product, int amount)
{
	auto it = this->productInitialAmountMap.find(product);
	if (it == this->productInitialAmountMap.end())
		this->productInitialAmountMap[product] = amount;
}

std::string SupplyManager::getHashCode()
{
	std::string thisAddress = std::to_string(reinterpret_cast<int>(this));
	std::string hashCode = std::to_string(std::hash<std::string>()(thisAddress));

	return hashCode;
}